Refactor gen_varkey_cmd

michael-grunder opened the pull request:
Use the new argument parsing API which lets us avoid an allocation and switch to redis_cmd_append_sstr_key_zval to abstract away key prefixing logic.

michael-grunder committed:
Refactor gen_varkey_cmd

Use the new argument parsing API which lets us avoid an allocation
and switch to redis_cmd_append_sstr_key_zval to abstract away key
prefixing logic.

yatsukhnenko commented on redis_commands.c:
```diff
@@ -1818,141 +1818,82 @@ int redis_key_str_arr_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
-        efree(z_args);
-        return FAILURE;
-    }
+    single_array = argc == 1 + !!has_timeout && Z_TYPE(argv[0]) == IS_ARRAY;
```
I can't parse this line 😄 Could you add parenthesis?

michael-grunder replied:
It made sense to me at the time.  Pushed updated logic :joy:

michael-grunder committed:
Clearer logic for checking if we have a single array.

michael-grunder committed:
Parentheses.

michael-grunder commented:
@yatsukhnenko You want any more changes to this?

yatsukhnenko commented:
I'm going to take a look one more time (when electricity restored 🥲)

michael-grunder committed:
Remove min_argc indirection all together.

Given how `gen_varkey_cmd` is called, we can actually derive the
`min_argc` from whether or not it has a timeout.

yatsukhnenko commented on redis_commands.c:
```diff
@@ -1818,141 +1818,85 @@ int redis_key_str_arr_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
+            ZVAL_COPY_VALUE(&ztimeout, &argv[argc - 1]);
+
+        if (Z_TYPE(ztimeout) != IS_LONG && Z_TYPE(ztimeout) != IS_DOUBLE) {
+            php_error_docref(NULL, E_WARNING, "Timeout must be a long or double");
```
Don't we need `zval_dtor(&ztimeout)` here?

michael-grunder replied:
No, `ZVAL_COPY_VALUE` does not increment the ref count.  It's a shallow copy.

I double-checked to be sure.  If I change `ZVAL_COPY_VALUE` -> `ZVAL_COPY`, PHP will report a leak.
